node/avn-service/src/ethereum_events_handler.rs

```diff
@@ -393,12 +394,107 @@ where
         + ApiExt<Block>
         + BlockBuilder<Block>,
 {
-    let (range, partition_id) = config
+    let result = &config
         .client
         .runtime_api()
         .query_active_block_range(config.client.info().best_hash)
-        .map_err(|err| format!("Failed to query active block range: {:?}", err))?;
+        .map_err(|err| format!("Failed to query bridge contract: {:?}", err))?;
+
+    match result {
+        // A range is active, attempt processing
+        Some((range, partition_id)) => {
+            process_events(
+                &config,
+                range.clone(),
+                *partition_id,
+                &current_public_key,
+                &events_registry,
+            )
+            .await?;
+        },
+        // There is no active range, attempt initial range voting.
+        None => {
+            submit_latest_ethereum_block(&config, &current_public_key).await?;
+        },
+    };
+    Ok(())
+}
 
+async fn submit_latest_ethereum_block<Block, ClientT>(
+    config: &EthEventHandlerConfig<Block, ClientT>,
+    current_public_key: &sp_core::sr25519::Public,
+) -> Result<(), String>
+where
+    Block: BlockT,
+    ClientT: BlockBackend<Block>
+        + UsageProvider<Block>
+        + HeaderBackend<Block>
+        + sp_api::ProvideRuntimeApi<Block>,
+    ClientT::Api: pallet_eth_bridge_runtime_api::EthEventHandlerApi<Block, AccountId>
+        + ApiExt<Block>
+        + BlockBuilder<Block>,
+{
+    let has_casted_vote = config
+        .client
+        .runtime_api()
+        .query_has_author_casted_vote(config.client.info().best_hash, current_public_key.0.into())
+        .map_err(|err| format!("Failed to check if author has casted latest  vote: {:?}", err))?;
+
+    if !has_casted_vote {
+        let web3_data_mutex = config.web3_data_mutex.lock().await;
+        let web3_ref = match web3_data_mutex.web3.as_ref() {
+            Some(web3) => web3,
+            None => return Err("Web3 connection not set up".into()),
+        };
+
+        let latest_seen_ethereum_block = web3_utils::get_current_block_number(web3_ref)
```

Comment:
Should we get `current_block` - X to take into account finality?

Reply:
I think there is no need because the value gets transformed into a range after reducing an amount of blocks, ensuring finality like that.
https://github.com/AventusProtocolFoundation/avn-parachain/blob/feat/SYS-3560-add-event-listener/primitives/avn-common/src/event_discovery.rs#L171-L179